metrics-tracing-context/src/tracing_integration.rs

```diff
@@ -146,6 +153,18 @@ where
         span.extensions_mut().insert(labels);
     }
 
+    fn on_record(&self, id: &Id, values: &Record<'_>, cx: Context<'_, S>) {
+        let span = cx.span(id).expect("span must already exist!");
+        let labels = Labels::from_record(values);
+
+        let ext = &mut span.extensions_mut();
+        if let Some(existing) = ext.get_mut::<Labels>() {
+            existing.extend_from_labels(&labels);
+        } else {
+            ext.insert(labels);
+        }
+    }
+
     unsafe fn downcast_raw(&self, id: TypeId) -> Option<*const ()> {
```

Comment:
By the way, there is a way to get rid of this `unsafe` implementation. If you're interested, I can provide more info, or even make a PR.

Reply:
There's not, as it's part of the [`Subscriber`](https://docs.rs/tracing-core/latest/tracing_core/trait.Subscriber.html#method.downcast_raw) trait itself.

Reply:
What I mean is it's not needed for functionality provided by this crate. There is, and I already used this.

Here's the general idea:

```rust
use tracing::{span, Dispatch, Subscriber};
use tracing_subscriber::{
    layer::SubscriberExt as _, registry::LookupSpan, util::SubscriberInitExt as _, Layer,
};

fn main() {
    tracing_subscriber::registry()
        .with(MySubscriber::new())
        .init();

    tracing::info_span!("parent").in_scope(|| {
        tracing::info_span!("child").in_scope(|| {
            assert_eq!(get_parent_name().as_deref(), Some("parent"));
        });
    });
}

fn get_parent_name() -> Option<String> {
    tracing::Span::current()
        .with_subscriber(|(id, dispatch)| {
            dispatch
                .downcast_ref::<MySubscriber>()?
                .get_parent_name(dispatch, id)
        })
        .flatten()
}

struct MySubscriber {
    get_parent_name: Option<fn(dispatch: &Dispatch, id: &span::Id) -> Option<String>>,
}

impl MySubscriber {
    fn new() -> Self {
        Self {
            get_parent_name: None,
        }
    }

    fn get_parent_name(&self, dispatch: &Dispatch, id: &span::Id) -> Option<String> {
        (self.get_parent_name?)(dispatch, id)
    }
}

impl<S> Layer<S> for MySubscriber
where
    S: Subscriber + for<'span> LookupSpan<'span>,
{
    fn on_layer(&mut self, _: &mut S) {
        self.get_parent_name = Some(|dispatch: &Dispatch, id: &span::Id| {
            let sub = dispatch.downcast_ref::<S>().unwrap();
            let span = sub.span(id).unwrap();

            span.parent().map(|p| p.name().to_owned())

            // Instead of this 👆, do some logic with callsite-provided data and spanref
        });
    }
}
```

Reply:
Hmm, interesting. Let me think about this, since I couldn't wrap my head around it on the first read-through. 😅

Reply:
I added helpful comments so it's easier to digest (hopefully)!

Reply:
Yeah, neat.

If you want to try reworking the current code to use this approach, in this same PR, I'd be fine with that... but also happy to review it in a separate PR so we can get this merged sooner.